File: src/sdk/ios/Sources/OFAAgent/Constraint/ConstraintChecker.swift
import Foundation

/// 约束类型
public struct ConstraintType: OptionSet, Sendable {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    public static let none = ConstraintType(rawValue: 0)
    public static let privacy = ConstraintType(rawValue: 1)
    public static let financial = ConstraintType(rawValue: 2)
    public static let security = ConstraintType(rawValue: 4)
    public static let authRequired = ConstraintType(rawValue: 8)
    public static let location = ConstraintType(rawValue: 16)
    public static let personal = ConstraintType(rawValue: 32)
    public static let device = ConstraintType(rawValue: 64)
}

/// 约束检查结果
public struct ConstraintResult: Sendable {
    public let allowed: Bool
    public let violated: ConstraintType
    public let reason: String?
    public let requiresAuth: Bool
    public var suggestions: [String]

    public init(allowed: Bool, violated: ConstraintType = .none, reason: String? = nil, requiresAuth: Bool = false) {
        self.allowed = allowed
        self.violated = violated
        self.reason = reason
        self.requiresAuth = requiresAuth
        self.suggestions = []
    }

    public mutating func addSuggestion(_ suggestion: String) {
        suggestions.append(suggestion)
    }
}

/// 约束规则
public struct ConstraintRule: Sendable {
    public let name: String
    public let type: ConstraintType
    public let actionPattern: String?
    public let dataPattern: String?
    public let offlineRestricted: Bool
    public let requiresAuth: Bool
    public let message: String

    public init(
        name: String,
        type: ConstraintType,
        actionPattern: String? = nil,
        dataPattern: String? = nil,
        offlineRestricted: Bool = false,
        requiresAuth: Bool = false,
        message: String = ""
    ) {
        self.name = name
        self.type = type
        self.actionPattern = actionPattern
        self.dataPattern = dataPattern
        self.offlineRestricted = offlineRestricted
        self.requiresAuth = requiresAuth
        self.message = message
    }
}

/// 约束检查器
public actor ConstraintChecker {
    private var rules: [ConstraintRule] = []
    private var offlineRestrictedActions: Set<String> = []
    private var sensitiveFields: Set<String> = []
    private var offlineMode = false

    public init() {
        loadDefaultRules()
        loadSensitiveFields()
    }

    private func loadDefaultRules() {
        // 财务操作
        addRule(ConstraintRule(
            name: "financial_operations",
            type: .financial,
            actionPattern: "(payment|transfer|withdraw|pay)",
            offlineRestricted: true,
            requiresAuth: true,
            message: "Financial operations require online mode and authorization"
        ))

        // 隐私数据
        addRule(ConstraintRule(
            name: "privacy_data",
            type: .privacy,
            dataPattern: "(idcard|id_card|身份证|passport|护照)",
            message: "Data contains sensitive personal information"
        ))

        // 位置信息
        addRule(ConstraintRule(
            name: "location_data",
            type: .location,
            dataPattern: "(location|gps|latitude|longitude|经纬度)",
            requiresAuth: true,
            message: "Location data sharing requires authorization"
        ))

        // 安全操作
        addRule(ConstraintRule(
            name: "security_operations",
            type: .security,
            actionPattern: "(delete|password|auth|login|logout)",
            offlineRestricted: true,
            requiresAuth: true,
            message: "Security operations require online mode and authorization"
        ))
    }

    private func loadSensitiveFields() {
        let fields = [
            "idcard", "id_card", "身份证", "passport", "护照",
            "phone", "mobile", "电话", "手机",
            "email", "邮箱",
            "address", "地址",
            "bank_account", "银行卡",
            "password", "密码",
            "token", "令牌",
            "secret", "密钥",
            "location", "gps", "位置"
        ]
        sensitiveFields = Set(fields.map { $0.lowercased() })
    }

    /// 添加规则
    public func addRule(_ rule: ConstraintRule) {
        rules.append(rule)

        if rule.offlineRestricted, let pattern = rule.actionPattern {
            let cleanPattern = pattern
                .replacingOccurrences(of: "(", with: "")
                .replacingOccurrences(of: ")", with: "")
            for action in cleanPattern.split(separator: "|") {
                offlineRestrictedActions.insert(action.trimmingCharacters(in: .whitespaces).lowercased())
            }
        }
    }

    /// 移除规则
    public func removeRule(_ name: String) {
        rules.removeAll { $0.name == name }
    }

    /// 设置离线模式
    public func setOfflineMode(_ offline: Bool) {
        offlineMode = offline
    }

    /// 检查约束
    public func check(action: String, data: String? = nil) -> ConstraintResult {
        var result = ConstraintResult(allowed: true)

        // 1. 检查离线受限操作
        if offlineMode {
            let actionLower = action.lowercased()
            for restricted in offlineRestrictedActions {
                if actionLower.contains(restricted) {
                    result = ConstraintResult(
                        allowed: false,
                        violated: [.financial, .security],
                        reason: "Action '\(action)' requires online mode",
                        requiresAuth: false
                    )
                    result.suggestions.append("Connect to network or use alternative offline action")
                    return result
                }
            }
        }

        // 2. 应用规则
        for rule in rules {
            let ruleResult = applyRule(rule, action: action, data: data)
            if !ruleResult.allowed {
                return ruleResult
            }
        }

        // 3. 检查敏感数据
        if let data = data {
            let dataResult = checkSensitiveData(data)
            if !dataResult.allowed {
                return dataResult
            }
        }

        return result
    }

    private func applyRule(_ rule: ConstraintRule, action: String, data: String?) -> ConstraintResult {
        var result = ConstraintResult(allowed: true)

        // 检查操作模式
        if let actionPattern = rule.actionPattern {
            if !action.range(of: actionPattern, options: .regularExpression)?.isEmpty ?? true {
                return result
            }
        }

        // 检查数据模式
        if let dataPattern = rule.dataPattern, let data = data {
            if !data.range(of: dataPattern, options: .regularExpression)?.isEmpty ?? true {
                return result
            }
        }

        // 离线限制
        if rule.offlineRestricted && offlineMode {
            return ConstraintResult(
                allowed: false,
                violated: rule.type,
                reason: rule.message,
                requiresAuth: rule.requiresAuth
            )
        }

        // 授权要求
        if rule.requiresAuth {
            // TODO: 检查用户授权状态
            return ConstraintResult(
                allowed: false,
                violated: .authRequired,
                reason: rule.message,
                requiresAuth: true
            )
        }

        return result
    }

    private func checkSensitiveData(_ data: String) -> ConstraintResult {
        let dataLower = data.lowercased()

        for field in sensitiveFields {
            if dataLower.contains(field) {
                var type: ConstraintType = .privacy
                var reason = "Data contains sensitive information"

                if field.contains("bank") || field.contains("card") {
                    type = .financial
                    reason = "Data contains financial information"
                } else if field.contains("location") || field.contains("gps") {
                    type = .location
                    reason = "Data contains location information"
                } else if field.contains("password") || field.contains("token") || field.contains("secret") {
                    type = .security
                    reason = "Data contains security credentials"
                }

                return ConstraintResult(
                    allowed: false,
                    violated: type,
                    reason: reason
                )
            }
        }

        return ConstraintResult(allowed: true)
    }

    /// 添加敏感字段
    public func addSensitiveField(_ field: String) {
        sensitiveFields.insert(field.lowercased())
    }

    /// 移除敏感字段
    public func removeSensitiveField(_ field: String) {
        sensitiveFields.remove(field.lowercased())
    }

    /// 获取离线受限操作
    public func getOfflineRestrictedActions() -> Set<String> {
        offlineRestrictedActions
    }

    /// 获取规则列表
    public func getRules() -> [ConstraintRule] {
        rules
    }

    /// 是否允许
    public func isAllowed(action: String, data: String? = nil) -> Bool {
        check(action: action, data: data).allowed
    }
}

File: src/sdk/android/src/main/java/com/ofa/agent/constraint/ConstraintChecker.java
package com.ofa.agent.constraint;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.regex.Pattern;

/**
 * 约束类型
 */
public enum ConstraintType {
    NONE(0),
    PRIVACY(1),
    FINANCIAL(2),
    SECURITY(4),
    AUTH_REQUIRED(8),
    LOCATION(16),
    PERSONAL(32),
    DEVICE(64);

    private final int value;

    ConstraintType(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }
}

/**
 * 约束检查结果
 */
public class ConstraintResult {
    public final boolean allowed;
    public final ConstraintType violated;
    public final String reason;
    public final boolean requiresAuth;
    public final List<String> suggestions;

    public ConstraintResult(boolean allowed, ConstraintType violated, String reason, boolean requiresAuth) {
        this.allowed = allowed;
        this.violated = violated;
        this.reason = reason;
        this.requiresAuth = requiresAuth;
        this.suggestions = new ArrayList<>();
    }

    public void addSuggestion(String suggestion) {
        suggestions.add(suggestion);
    }
}

/**
 * 约束规则
 */
public class ConstraintRule {
    public final String name;
    public final ConstraintType type;
    public final Pattern actionPattern;
    public final Pattern dataPattern;
    public final boolean offlineRestricted;
    public final boolean requiresAuth;
    public final String message;

    public ConstraintRule(String name, ConstraintType type, String actionPattern, String dataPattern,
                          boolean offlineRestricted, boolean requiresAuth, String message) {
        this.name = name;
        this.type = type;
        this.actionPattern = actionPattern != null ? Pattern.compile(actionPattern, Pattern.CASE_INSENSITIVE) : null;
        this.dataPattern = dataPattern != null ? Pattern.compile(dataPattern, Pattern.CASE_INSENSITIVE) : null;
        this.offlineRestricted = offlineRestricted;
        this.requiresAuth = requiresAuth;
        this.message = message;
    }
}

/**
 * 约束检查器
 */
public class ConstraintChecker {
    private static final String TAG = "ConstraintChecker";

    private final List<ConstraintRule> rules = new CopyOnWriteArrayList<>();
    private final Set<String> offlineRestrictedActions = new HashSet<>();
    private final Set<String> sensitiveFields = new HashSet<>();
    private boolean offlineMode = false;

    public ConstraintChecker() {
        loadDefaultRules();
        loadSensitiveFields();
    }

    private void loadDefaultRules() {
        // 财务操作
        addRule(new ConstraintRule(
            "financial_operations",
            ConstraintType.FINANCIAL,
            "(payment|transfer|withdraw|pay)",
            null,
            true,
            true,
            "Financial operations require online mode and authorization"
        ));

        // 隐私数据
        addRule(new ConstraintRule(
            "privacy_data",
            ConstraintType.PRIVACY,
            null,
            "(idcard|id_card|身份证|passport|护照)",
            false,
            false,
            "Data contains sensitive personal information"
        ));

        // 位置信息
        addRule(new ConstraintRule(
            "location_data",
            ConstraintType.LOCATION,
            null,
            "(location|gps|latitude|longitude|经纬度)",
            false,
            true,
            "Location data sharing requires authorization"
        ));

        // 安全操作
        addRule(new ConstraintRule(
            "security_operations",
            ConstraintType.SECURITY,
            "(delete|password|auth|login|logout)",
            null,
            true,
            true,
            "Security operations require online mode and authorization"
        ));
    }

    private void loadSensitiveFields() {
        String[] fields = {
            "idcard", "id_card", "身份证", "passport", "护照",
            "phone", "mobile", "电话", "手机",
            "email", "邮箱",
            "address", "地址",
            "bank_account", "银行卡",
            "password", "密码",
            "token", "令牌",
            "secret", "密钥",
            "location", "gps", "位置"
        };
        for (String field : fields) {
            sensitiveFields.add(field.toLowerCase());
        }
    }

    public void addRule(@NonNull ConstraintRule rule) {
        rules.add(rule);
        if (rule.offlineRestricted && rule.actionPattern != null) {
            // 提取动作关键词
            String pattern = rule.actionPattern.pattern();
            pattern = pattern.replace("(", "").replace(")", "");
            for (String action : pattern.split("\\|")) {
                offlineRestrictedActions.add(action.trim().toLowerCase());
            }
        }
    }

    public void removeRule(@NonNull String name) {
        rules.removeIf(rule -> rule.name.equals(name));
    }

    public void setOfflineMode(boolean offline) {
        this.offlineMode = offline;
    }

    @NonNull
    public ConstraintResult check(@NonNull String action, @Nullable String dataJson) {
        ConstraintResult result = new ConstraintResult(true, ConstraintType.NONE, null, false);

        // 1. 检查离线受限操作
        if (offlineMode) {
            String actionLower = action.toLowerCase();
            for (String restricted : offlineRestrictedActions) {
                if (actionLower.contains(restricted)) {
                    result = new ConstraintResult(false, ConstraintType.FINANCIAL,
                        "Action '" + action + "' requires online mode", false);
                    result.addSuggestion("Connect to network or use alternative offline action");
                    return result;
                }
            }
        }

        // 2. 应用规则
        for (ConstraintRule rule : rules) {
            ConstraintResult ruleResult = applyRule(rule, action, dataJson);
            if (!ruleResult.allowed) {
                return ruleResult;
            }
        }

        // 3. 检查敏感数据
        if (dataJson != null && !dataJson.isEmpty()) {
            ConstraintResult dataResult = checkSensitiveData(dataJson);
            if (!dataResult.allowed) {
                return dataResult;
            }
        }

        return result;
    }

    private ConstraintResult applyRule(ConstraintRule rule, String action, String dataJson) {
        ConstraintResult result = new ConstraintResult(true, ConstraintType.NONE, null, false);

        // 检查操作模式
        if (rule.actionPattern != null && !rule.actionPattern.matcher(action).find()) {
            return result;
        }

        // 检查数据模式
        if (rule.dataPattern != null && dataJson != null) {
            if (!rule.dataPattern.matcher(dataJson).find()) {
                return result;
            }
        }

        // 离线限制
        if (rule.offlineRestricted && offlineMode) {
            return new ConstraintResult(false, rule.type, rule.message, rule.requiresAuth);
        }

        // 授权要求
        if (rule.requiresAuth) {
            // TODO: 检查用户授权状态
            return new ConstraintResult(false, ConstraintType.AUTH_REQUIRED, rule.message, true);
        }

        return result;
    }

    private ConstraintResult checkSensitiveData(String dataJson) {
        String dataLower = dataJson.toLowerCase();

        for (String field : sensitiveFields) {
            if (dataLower.contains(field)) {
                ConstraintType type = ConstraintType.PRIVACY;
                String reason = "Data contains sensitive information";

                if (field.contains("bank") || field.contains("card")) {
                    type = ConstraintType.FINANCIAL;
                    reason = "Data contains financial information";
                } else if (field.contains("location") || field.contains("gps")) {
                    type = ConstraintType.LOCATION;
                    reason = "Data contains location information";
                } else if (field.contains("password") || field.contains("token") || field.contains("secret")) {
                    type = ConstraintType.SECURITY;
                    reason = "Data contains security credentials";
                }

                return new ConstraintResult(false, type, reason, false);
            }
        }

        return new ConstraintResult(true, ConstraintType.NONE, null, false);
    }

    public void addSensitiveField(@NonNull String field) {
        sensitiveFields.add(field.toLowerCase());
    }

    public void removeSensitiveField(@NonNull String field) {
        sensitiveFields.remove(field.toLowerCase());
    }

    @NonNull
    public Set<String> getOfflineRestrictedActions() {
        return new HashSet<>(offlineRestrictedActions);
    }

    @NonNull
    public List<ConstraintRule> getRules() {
        return new ArrayList<>(rules);
    }
}